### DISCUSSION TITLE
Support separate failing and succeeding conditions for alerts

### DISCUSSION THREAD

**Discussion by leo60228** [grafana/grafana#47181] [Alerting]
**What would you like to be added**: Support for specifying separate failing and succeeding conditions for alerts. For example, an alert could trigger when a value goes above 100, and then go back to normal once it goes below 90. This is similar to #6202, but that was ambiguous between this and debouncing.

**Why is this needed**: This would allow further control over some types of flapping alerts. Debouncing is able to prevent an alert for going above 100 from repeatedly triggering for this example:
![image](https://user-images.githubusercontent.com/8355305/103925522-ab80a600-50e5-11eb-8937-50846486482a.png)

However, this allows preventing this from repeatedly triggering:
![image](https://user-images.githubusercontent.com/8355305/103925566-be937600-50e5-11eb-8380-8c0738ca3dd3.png)

While in many cases one solution will work for both, a combination of the two is necessary to cover all cases.

**Comment by torkelo** [comment]
The existing For parameter should solve this I think? @kylebrandt any thoughts?

**Comment by kylebrandt** [comment]
@torkelo I believe `FOR` is only a hold for going into a bad state. I do not know what Debouncing is.

For alerting NG, I have had the idea of "Recovery". Which is like FOR, but is hold for going from an Alerting State back to a normal State (requires X Normal events, or X amount of time with Normal events), before going back to State normal. For my spec:

> Recovery
> Many systems do not have a notion of recovery - rather they return to normal state. Well this is simpler, a recovery condition can be very useful. This is a condition that is required for the alert instance to return back to normal from a previously non-normal state. By adding the possibility of recovery, a user can make alerts that are more sensitive to trigger, while also being less sensitive to go back to normal, which can reduce alert noise -- asymmetrical alert conditions.

But I haven't gotten to designing the state transition table for the Recovery feature/state yet (was planning on doing that today already actually), so I am not sure how it will work out. These holding state features can be tricky - need to make sure there that are not scenarios where you end up flapping between various "bad" conditions and never sent a notification. So all the transitions need to be mapped out.

Another option would be to support specifying a recovery condition that is different from the trigger condition. Such that it only goes out of Alerting if the alerting condition is not true and the recovery condition is true. But I think that creates more user facing complexity in what they need to input and am leaning towards having Recovery like FOR, but flipped.

**Comment by leo60228** [comment]
Debouncing is when a change in state that only occurs for a short time is ignored (`FOR` in Grafana). The term originated from working around the problem that pushbuttons work by having the user press two contacts together, and those contacts would literally "bounce" leading to several short pulses on and off before reaching a steady state. I guess the term is more field-specific than I thought, sorry about that.

**Comment by leo60228** [comment]
Here's a quick mockup of the interface I'm picturing:
![image](https://user-images.githubusercontent.com/8355305/104036730-c4e72800-51a1-11eb-8a00-c4676da83355.png)

**Comment by kylebrandt** [comment]
Anything we do along these lines I think will be in Alerting NG, do not really have enough to show right now on that (but beta in 8, and will exist side by side with current alerting. See https://github.com/grafana/grafana/issues/6983#issuecomment-712915673 and https://github.com/grafana/grafana/issues/7832#issuecomment-712936681 for some things about alerting ng). But your screenshot gets the idea across. There are two main things in it:

 1. An additional recovery query to run (I'm going to call this the condition)
 2. An hold timer (For, but a different one for recovery since it can have a different duration).

Adding the additional query and not just a timer for recovery is more powerful, but also more complex. As there is a (potentially) different query to run in order for recovery.

So I think we may have separate recovery FOR timer, but it will be tied to the same condition (at least in earlier versions) in order to keep the complexity down (which there is a potentially a lot of, as mentioned in https://github.com/grafana/grafana/issues/7832#issuecomment-712963426).

**Comment by caroso-de** [comment]
I'm so glad I found this issue. Imho, Grafana definitely needs hysteresis on alerts

I've missed it on several occasions already. The reason is that no matter how much you smooth out your data, you cannot prevent your alerting condition from constantly flipping, once the values come very close to your threshold. 

For example, I have a temperature sensor which slowly reached its alerting threshold yesterday and since then makes the condition flip every 30 Minutes or so.

  
Btw, the usual approach with hysteresis is simply to assume a different threshold on exactly the same measurement, e.g. if your alert-condition is "avg(A,15m,now) > 100" you'd simply use <90 or something like that as the recovery condition. 

I think the recovery condition will work perfectly fine with the other parameters (debouncing, averaging) being exactly the same (logic inverted) because these parameters depend mainly on how your measurements behave. 


tl;dr this issue is highly underrated, implementing this will be a huge step forward

And thanks for a great tool,

a happy grafana user

**Comment by JohnnyQQQQ** [comment]
As this is a feature request, we will move it to discussions for now till we move forward with it.

**Comment by jantman** [comment]
I'm not sure if there's been any progress on this, but it's an extremely important feature for me. My meta use case is that there are things where I want to be notified ASAP if a threshold is crossed (i.e. very low `every` and `FOR`) but not notified many, many times in rapid succession if the value is flapping between firing and resolved.

In a more specific example, here's a graph of some really spikey IO on a NAS disk, where it was essentially oscillating between complete IO saturation (99-100%) and "normal" low IO:

![synology-nas-details-snmp-14284](https://user-images.githubusercontent.com/273454/217092822-0ad7cebc-c5fa-434d-b390-4aaef21b4277.png)

I absolutely do want to be notified ASAP (in my case, "every 1 minute for 1 minute") if the disk is at IO capacity. I do not, however, want to get *lots and lots* of notifications during this event.

A separate recovery `FOR` timer, as suggested previously, would be a simple (at least from the UX side) solution that would definitely solve my use case.

**Comment by yuri-tceretian** [comment]
I think a new "recovery threshold" feature proposed in PR https://github.com/grafana/grafana/pull/75189 might be helpful in this case too.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). We are migrating this discussion to an issue and closing the discussion. The issue is #82954. Feel free to continue the discussion around this there. Thank you!
